plugins/performance-lab/includes/admin/plugins.php

```diff
@@ -54,61 +54,94 @@ function perflab_query_plugin_info( string $plugin_slug ) {
+			),
+		);
+
+		$has_errors = true;
 	}
 
-	$plugins      = array();
-	$plugin_queue = perflab_get_standalone_plugins();
+	if ( ! $has_errors && is_object( $response ) && property_exists( $response, 'plugins' ) ) {
+		$plugin_queue = perflab_get_standalone_plugins();
 
-	// Index the plugins from the API response by their slug for efficient lookup.
-	$all_performance_plugins = array_column( $response->plugins, null, 'slug' );
+		// Index the plugins from the API response by their slug for efficient lookup.
+		$all_performance_plugins = array_column( $response->plugins, null, 'slug' );
 
-	// Start processing the plugins using a queue-based approach.
-	while ( count( $plugin_queue ) > 0 ) { // phpcs:ignore Squiz.PHP.DisallowSizeFunctionsInLoops.Found
-		$current_plugin_slug = array_shift( $plugin_queue );
+		// Start processing the plugins using a queue-based approach.
+		while ( count( $plugin_queue ) > 0 ) { // phpcs:ignore Squiz.PHP.DisallowSizeFunctionsInLoops.Found
+			$current_plugin_slug = array_shift( $plugin_queue );
 
-		if ( isset( $plugins[ $current_plugin_slug ] ) ) {
-			continue;
-		}
+			if ( isset( $plugins[ $current_plugin_slug ] ) ) {
+				continue;
+			}
 
-		if ( ! isset( $all_performance_plugins[ $current_plugin_slug ] ) ) {
-			return new WP_Error(
-				'plugin_not_found',
-				__( 'Plugin not found in WordPress.org API response.', 'performance-lab' )
-			);
+			if ( ! isset( $all_performance_plugins[ $current_plugin_slug ] ) ) {
+				// Cache the fact that the plugin was not found.
+				$plugins[ $current_plugin_slug ] = array(
+					'error' => array(
+						'code'    => 'plugin_not_found',
+						'message' => __( 'Plugin not found in API response.', 'performance-lab' ),
+					),
+				);
+				continue;
+			}
+
+			$plugin_data                     = $all_performance_plugins[ $current_plugin_slug ];
+			$plugins[ $current_plugin_slug ] = wp_array_slice_assoc( $plugin_data, $fields );
+
+			// Enqueue the required plugins slug by adding it to the queue.
+			if ( isset( $plugin_data['requires_plugins'] ) && is_array( $plugin_data['requires_plugins'] ) ) {
+				$plugin_queue = array_merge( $plugin_queue, $plugin_data['requires_plugins'] );
+			}
 		}
 
-		$plugin_data                     = $all_performance_plugins[ $current_plugin_slug ];
-		$plugins[ $current_plugin_slug ] = wp_array_slice_assoc( $plugin_data, $fields );
+		if ( ! isset( $plugins[ $plugin_slug ] ) ) {
+			// Cache the fact that the plugin was not found.
+			$plugins[ $plugin_slug ] = array(
+				'error' => array(
+					'code'    => 'plugin_not_found',
+					'message' => __( 'Plugin not found in API response.', 'performance-lab' ),
+				),
+			);
 
-		// Enqueue the required plugins slug by adding it to the queue.
-		if ( isset( $plugin_data['requires_plugins'] ) && is_array( $plugin_data['requires_plugins'] ) ) {
-			$plugin_queue = array_merge( $plugin_queue, $plugin_data['requires_plugins'] );
+			$has_errors = true;
 		}
```

Comment:
Now that I look at this again, what is this particular check needed for? Wouldn't this already be catered for by the check above in line 101?

Reply:
This check is needed because if the `$plugin_slug` is not in the `perflab_get_standalone_plugins()` then this `$plugin_slug` will not be processed inside the loop.

Reply:
In that case, I think we should make this check at the very beginning of the function, even before looking at the transient cache. This function should never be used for a plugin that is not among our plugins I think, so we could already check that before all the main logic in this function.

We wouldn't want that someone calling the function with e.g. `jetpack` would trigger this API request and logic to run for every single page load just because `jetpack` would (rightfully) never be included in the response.

Reply:
If we only check the `$plugin_slug` inside `perflab_get_standalone_plugins()` it will also cause error for the dependency plugins. And if we want to dynamically check the dependency then we need request the API.

Reply:
Why does the function need to be called for dependency plugins?

Reply:
The `perflab_render_plugin_card()` function is calling the `perflab_get_plugin_availability()`  function
and inside the `perflab_get_plugin_availability()` on [here](https://github.com/WordPress/performance/blob/69eca9a7936701475f84a0cabbfec36e1cb3fe27/plugins/performance-lab/includes/admin/plugins.php#L310C2-L320C3) it is calling the `perflab_query_plugin_info()` for the dependency plugin.

Reply:
Ah, that's a good point. But in that case, I think at least this error should use a different message? I would try to clarify it as: If the plugin is not found at this point, we know it's not one of our plugins, since it's not in the API response, and the API response should always included all our plugins _plus_ its dependencies.